mrmustard/physics/gaussian_integrals.py

```diff
@@ -472,3 +392,112 @@ def contract_two_Abc_poly(
     )
+
+
+def complex_gaussian_integral_2(
+    Abc1: tuple,
+    Abc2: tuple,
+    idx1: Sequence[int],
+    idx2: Sequence[int],
+    measure: float = -1,
+    batched: bool = False,
+):
+    r"""Computes the (optionally batched) Gaussian integral of the product of two exponential functions of complex quadratic forms with
+    respect to an exponential measure. The first dimension of the ``Abc`` parameters is batched over. The ``c`` parameter
+    is allowed to have additional dimensions which are to be combined using an outer product.
+    When ``batched=True``, this function expects ``A1.shape=(b,n1,n1)``, ``A2.shape=(b,n2,n2)``, ``b1.shape=(b,n1)``, ``b2.shape=(b,n2)``,
+    ``c1.shape=(b,d1*)``, ``c2.shape=(b,d2*)``. The output is the A,b,c parametrization of the exponential of a quadratic form
+    with parameters of shape ``A.shape=(b,n1+n2-2m,n1+n2-2m)``, ``b.shape=(b,n1+n2-2m)``, ``c.shape=(b,d1*,d2*)``.
+    When ``batched=False``, this function expects ``A1.shape=(n1,n1)``, ``A2.shape=(n2,n2)``, ``b1.shape=(n1,)``, ``b2.shape=(n2,)``,
+    ``c1.shape=(d1*)``, ``c2.shape=(d2*)``. The output is the A,b,c parametrization of the exponential of a quadratic form
+    with parameters of shape ``A.shape=(n1+n2-2m,n1+n2-2m)``, ``b.shape=(n1+n2-2m)``, ``c.shape=(d1*,d2*)``.
+
+    The integral is defined as
+
+    :math:`\int_{C^m} F_1(z)F_2(z^*) d\mu(z)`
+
+    where the functions ``F_1`` and ``F_2`` are defined over ``n_1`` and ``n_2`` variables respectively, and the integral is taken
+    over a subset of  ``m`` pairs of variables whose position is indicated by ``idx1`` and ``idx2``.
+    Therefore the resulting function is a new exponential of a quadratic form defined over ``n_1 + n_2 - 2m`` variables.
+
+    The functions are assumed to be in the form
+    :math:`F_i(z) = c_i\textrm{exp}(0.5 z^T A_i z + b_i^T z),\quad z\in\mathbb{C}^{n_i}`
+
+    where ``A_i`` is complex symmetric
+
+    The integration measure is given by (``s`` here is the ``measure`` argument):
+    :math:`dmu(z) = \textrm{exp}(s * |z|^2) \frac{d^{2n}z}{\pi^n} = \frac{1}{\pi^n}\textrm{exp}(s * |z|^2) d\textrm{Re}(z) d\textrm{Im}(z)`
+
+    Arguments:
+        A1,b1,c1: the ``(A,b,c)`` triple that defines ``F_1(z)``
+        A2,b2,c2: the ``(A,b,c)`` triple that defines ``F_2(z^*)``
+        idx1: the tuple of indices indicating which variables of F_1 to integrate over
+        idx2: the tuple of indices indicating which variables of F_2 to integrate over
+        measure: the exponent of the measure (default is -1: Bargmann measure)
+
+    Returns:
+        The ``(A,b,c)`` triple of the result of the integral.
+
+    Raises:
+        ValueError: If ``idx1`` and ``idx2`` have different lengths or if A_i, b_i and c_i have non-matching batch size.
+    """
+    # first we ensure that Ai, bi, ci have the correct shapes
+    A1, b1, c1 = Abc1
+    A2, b2, c2 = Abc2
+    c1 = math.astensor(c1)
+    c2 = math.astensor(c2)
```

Comment:
Can't we assume that the cs have already been converted to our backend? It looks like we're already making this assumption with the As and bs since we don't call `astensor` for those

Reply:
the problem is that c's could be simple scalars with no shape to them